background/index.js

```diff
@@ -17,13 +17,171 @@
 /* global chrome */
 
 import Processor from './processor';
+import Ws from './ws';
 
-const processor = new Processor();
+import { UI, TRANSPORT_UNINITIALIZED } from '../shared';
 
 chrome.runtime.onConnect.addListener((port) => {
-  console.assert(port.name === 'id');
+  if (port.name === 'secureApi') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'barScripts') {
+    port.onMessage.addListener(loadScripts(port));
+    return;
+  }
+
+  // TODO [ToDr] Use connection without elevated privileges here!
+  if (port.name === 'web3') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'id') {
+    port.onMessage.addListener(processId(port));
+    return;
+  }
+
+  throw new Error(`Unrecognized port: ${port.name}`);
+});
+
+let openedTabId = null;
+let transport = null;
+// Attempt to extract token on start if not available.
+extractToken();
+
+chrome.runtime.onMessage.addListener((request, sender, callback) => {
+  if (!(transport && transport.isConnected) && request.token) {
+    if (transport) {
+      // TODO [ToDr] kill old transport!
+    }
+
+    if (openedTabId) {
+      chrome.tabs.remove(openedTabId);
+      openedTabId = null;
+    }
+
+    console.log('Extracted a token: ', request.token);
+    console.log('Extracted backgroundSeed: ', request.backgroundSeed);
+    chrome.storage.local.set({
+      'authToken': request.token,
+      'backgroundSeed': request.backgroundSeed
+    }, () => {});
+    transport = new Ws(`ws://${UI}`, request.token, true);
+    return;
+  }
+});
+
+let codeCache = null;
+function loadScripts (port) {
+  function retry (msg) {
+    if (msg.type !== 'parity.bar.code') {
+      return;
+    }
+
+    if (!codeCache) {
+      const vendor = fetch(`http://${UI}/vendor.js`)
+        .then(x => x.blob());
+      const embed = fetch(`http://${UI}/embed.html`)
+        .then(x => x.text())
+        .then(page => ({
+          styles: /styles\/embed\.([a-z0-9]{10})\.css/.exec(page),
+          scripts: /embed\.([a-z0-9]{10})\.js/.exec(page)
+        }))
+        .then(res => {
```

Comment:
We might want to use a ServiceWorker here. It's its job to fetch and cache requests, and it's work pretty good. Also it might be good to have a way to export which files are required by the `embed.html` file, instead of using some RegExp. There might be a Webpack plugin/option for that